From 643712ca70b924c0edcc059699aa1ee42234be34 Mon Sep 17 00:00:00 2001
From: John MacFarlane <jgm@berkeley.edu>
Date: Fri, 19 Sep 2025 10:15:32 +0200
Subject: [PATCH] LaTeX writer: avoid `\_` in bibliography variable.

Since underscores are common in filenames, and pandoc will
render strings to variables using default LaTeX escaping, we
special-case `bibliography`, under the assumption that this variable
will be used in the context of the `\bibliography{..}` command,
which accepts unescaped underscores.

Closes #11152.
---
 src/Text/Pandoc/Writers/LaTeX.hs | 10 +++++++++-
 test/djot-reader.native          |  2 +-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/Text/Pandoc/Writers/LaTeX.hs b/src/Text/Pandoc/Writers/LaTeX.hs
index 24ed4bebc84e..14c8de08a160 100644
--- a/src/Text/Pandoc/Writers/LaTeX.hs
+++ b/src/Text/Pandoc/Writers/LaTeX.hs
@@ -215,7 +215,15 @@ pandocToLaTeX options (Pandoc meta blocks) = do
                            _         -> [])
                     $ lookupMetaInlines "nocite" meta
 
-  let context  =  defField "toc" (writerTableOfContents options) $
+   -- see #7414, avoid escaped underscores
+  let unescapeUnderscore = T.replace "\\_" "_"
+  let bibliography' = map unescapeUnderscore <$>
+                        getField "bibliography" metadata
+
+  let context  =  (case bibliography' of
+                     Nothing -> id
+                     Just xs -> resetField "bibliography" xs) $
+                  defField "toc" (writerTableOfContents options) $
                   defField "lof" (writerListOfFigures options) $
                   defField "lot" (writerListOfTables options) $
                   defField "toc-depth" (tshow
diff --git a/test/djot-reader.native b/test/djot-reader.native
index edd1323a534a..bd0ae351ff9a 100644
--- a/test/djot-reader.native
+++ b/test/djot-reader.native
@@ -98,7 +98,7 @@ Pandoc
           [ Para [ Str "Code in a block quote:" ]
           , CodeBlock
               ( "" , [ "" ] , [] )
-              "sub status {\nprint \"working\";\n}\n"
+              "sub status {\n    print \"working\";\n}\n"
           , Para [ Str "A list:" ]
           , OrderedList
               ( 1 , Decimal , Period )

